### PR TITLE
Sandrine's Cluedo mystery picker

### PR DESCRIPTION
**Pull request by MindstormingAB**
For this Cluedo project, I first followed the different steps described in the brief to reach the blue level. Thanks to the hints left in the original code, it felt pretty clear so I decided to challenge myself with red and black levels.

I struggled a lot with scope and bad initialization timing but, at then end, I managed to solve everything and I'm satisfied with the final result:
- responsive design
- a shuffleFavouriteWeapon function
- loading animations
- changing the odd of the weapon and the room randomizers
- displaying the information in the DOM
- saving all randomized information in the mystery object.

Link to the deployed project: https://sandrinescluedo.netlify.app/

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,22 @@
...
 
 ## The problem
 
-Describe how you approached to problem, and what tools and techniques you used to solve it. How did you plan? What technologies did you use? If you had more time, what would be next?
+I followed the different steps described in the brief. I was positively surprised to find so many hints in the original code and I found it quite easy to achieve the blue level (at least before styling).
+
+After Wednesday's lecture with Van, it felt pretty clear how to build the shuffleFavouriteWeapon function but it required a little bit more time to make the loading animations works as I attended.
+
+And then I decided to go for the black level... I spent my Saturday on it, facing several issues related to scope and bad initialization timing, as illustrated in my StackOverflow question (https://stackoverflow.com/c/technigo/questions/1545). After hours of restructuring and control.log, I managed to manipulate the odd in the logic behind generating weapon. Then it was much easier to replicate the solution on rooms. I still don't get why we had to put an id in the weapon array though...
```
Comment (sofiavazs) [inline README.md:12]: You did a great job! Awesome!👏🏻

File: code/clue.js [hunk 2]
```diff
@@ -1,59 +1,464 @@
...
+const roomName = document.getElementById('roomName');
+const roomImage = document.getElementById('roomImage');
+const revealedMystery = document.getElementById('mystery');
+const revealButton = document.getElementById('revealButton');
+const resetButton = document.getElementById('resetButton');
+const shuffleButton = document.getElementById('shuffleButton');
+const killerLoader = document.getElementById('killerLoader');
+const weaponLoader = document.getElementById('weaponLoader');
+const roomLoader = document.getElementById('roomLoader');
+
```
Comment (sofiavazs) [inline code/clue.js:283]: Nice way of structuring by organising all the variables in one place! Need to try that myself! 😊

File: code/clue.js [hunk 3]
```diff
@@ -1,59 +1,464 @@
...
+    revealedMystery.innerHTML = ('Pick a killer first!');
+  } else {
+    weaponLoader.style.display = "block";
+    weaponName.innerHTML = '';
+    weaponWeight.innerHTML = '';
+    weaponImage.src = '';
+    weaponImage.style.display = "none";
+    weaponFavouriteRoom.innerHTML = '';
+    revealedMystery.innerHTML = '';
+    setTimeout(function () {
```
Comment (sofiavazs) [inline code/clue.js]: oh no! A `function` 🚫! Arrow functions are the way => {😬}

File: code/clue.js [hunk 4]
```diff
@@ -1,59 +1,464 @@
...
+  if (mystery.killer === undefined) {
+    revealedMystery.innerHTML = ('Pick a killer first!');
+  } else if (mystery.weapon === undefined) {
+    revealedMystery.innerHTML = ('Pick a weapon first!');
+  } else {
+    roomLoader.style.display = "block";
+    roomName.innerHTML = '';
+    roomImage.src = '';
+    roomImage.style.display = "none";
+    setTimeout(function () {
```
Comment (sofiavazs) [inline code/clue.js]: same as above regarding `function` 🙈

File: code/clue.js [hunk 5]
```diff
@@ -1,59 +1,464 @@
...
+  ];
+  console.log(favouriteRoomDuplicated(weapon.favouriteRoom, 30));
+  console.log(randomRooms);
+  mystery.pickedRoom = randomSelector(randomRooms);
+  const mysteryRoom = rooms.find((item) => {
+    if (item.name === mystery.pickedRoom) {
+      return item;
+    };
+  });
+  console.log(mysteryRoom);
```
Comment (sofiavazs) [inline code/clue.js]: console.logs are awesome in development but not to be left behind 😇

File: code/clue.js [hunk 6]
```diff
@@ -1,59 +1,464 @@
...
+
+// Functions related to revelation of mystery 
+
+const revealMystery = () => {
+  if (mystery.killer === undefined || mystery.weapon === undefined || mystery.room === undefined) {
+    revealedMystery.innerHTML = ('Make sure you have picked a killer, a weapon and a room!');
+  } else {
+    revealedMystery.innerHTML = (`The murder was committed by ${mystery.killer.firstName} ${mystery.killer.lastName} with the ${mystery.weapon.name} in the ${mystery.room.name}.`);
+  }
+};
```
Comment (sofiavazs) [inline code/clue.js]: Really nice that you made sure the user gets prompted to pick cards and to pick the killer first! 😎

File: code/clue.js [hunk 7]
```diff
@@ -1,59 +1,464 @@
...
+
+const resetMystery = () => {
+  location.reload();
+};
```
Comment (sofiavazs) [inline code/clue.js:455]: Nice touch with the "reset" button as well!

File: code/index.html [hunk 8]
```diff
@@ -1,37 +1,87 @@
...
+  <header>
+    <h1>Cluedo mystery picker</h1>
+  </header>
+  <main>
     <section class="cards">
       <article class="killer-deck" id="killerCard">
-        <p>KILLER</p>
+        <p class="card-headers">KILLER</p>
+        <div class="loader" id="killerLoader">
+          <i class="fa fa-circle-o-notch fa-spin"></i>
```
Comment (sofiavazs) [inline code/index.html:24]: Great that you used an icon tag for the loader!

File: code/index.html [hunk 9]
```diff
@@ -1,37 +1,87 @@
...
-    <title>Technigo Project Cluedo</title>
-  </head>
-  <body>
+
+<head>
+  <meta charset="UTF-8" />
+  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+  <meta http-equiv="X-UA-Compatible" content="ie=edge" />
+  <link rel="stylesheet" href="style.css" />
+  <title>🔍 Sandrine's Cluedo</title>
```
Comment (sofiavazs) [inline code/index.html:9]: Cool touch! 🕵️‍♀️

File: code/clue.js [hunk 10]
```diff
@@ -1,59 +1,464 @@
...
+weaponCard.onclick = favouriteWeaponAnimation;
+
+// Actions related to Room deck
+
+const favouriteRoomAnimation = () => {
+  if (mystery.killer === undefined) {
+    revealedMystery.innerHTML = ('Pick a killer first!');
+  } else if (mystery.weapon === undefined) {
+    revealedMystery.innerHTML = ('Pick a weapon first!');
+  } else {
```
Comment (sofiavazs) [inline code/clue.js]: Awesome job with the user text message prompts! 👏🏻

**Review by sofiavazs — CHANGES_REQUESTED** [review]
I just want to say that you did such a great job Sandrine! Your project looks great and works as intended.
The code is easily readable, organised and nicely structured it was great to review it and I learned a lot too!
I loved the way you added these little touches throughout the project, like little messages as prompts to the user, the reset and the shuffle favourite weapon button, as well as, making the whole page responsive! Really impressive! 👏🏻
I can tell that you've put a lot of effort to achieve the stretch goals and that doesn't go unnoticed, keep it up!  💪

Super tiny things to fix: 
- `function` to an arrow function. example: `const this = () => {}`
-  Remove console.logs from code 

These are really tiny things but I think it's important to highlight them (since they're in our code guidelines), otherwise everything looked really impressive! 👏🏻

File: code/clue.js (hunk 3 above)
Comment (MindstormingAB) [reply code/clue.js]: @sofiavazs I found a description of the setTimeOut method on w3schools (https://www.w3schools.com/jsref/met_win_settimeout.asp) and I did not dare to change it 😬. But now, I did it, thanks to your comment! 💐0

File: code/clue.js (hunk 5 above)
Comment (MindstormingAB) [reply code/clue.js]: Argh! I deleted the ones I commented out and left the visible ones! Thanks for pointing that out, I'm gonna take care of them 🚮

File: code/clue.js (hunk 7 above)
Comment (MindstormingAB) [reply code/clue.js:455]: I hope this lazy solution will be accepted by the code teacher 😎

File: code/index.html (hunk 8 above)
Comment (MindstormingAB) [reply code/index.html:24]: Kind of a quick fix but I can recommend it, it is very easy to use. I first used padding to place it in the middle of the card and it spun in a very comical way 😂, margin works better.

**Review by sofiavazs — APPROVED** [review]

File: code/clue.js (hunk 6 above)
Comment (HIPPIEKICK) [reply code/clue.js]: Could also be written the other way around:
```
if (mystery.killer && mystery.weapon && mystery.room) {
  //Reveal the mystery
} else {
  //Pick some cards first
}
```

File: code/clue.js (hunk 7 above)
Comment (HIPPIEKICK) [reply code/clue.js:455]: Lazy is easy ;)

File: code/clue.js (hunk 10 above)
Comment (HIPPIEKICK) [reply code/clue.js]: This could be written like:
`if (!mystery.killer)`

File: code/clue.js [hunk 11]
```diff
@@ -1,59 +1,460 @@
...
+  candlestick,
+  dumbbell,
+  poison,
+  axe,
+  bat,
+  trophy,
+  pistol
+]
+
+const rooms = [
```
Comment (HIPPIEKICK) [inline code/clue.js:243]: Super nice with some room objects as well! 💯

File: code/clue.js [hunk 12]
```diff
@@ -1,59 +1,460 @@
...
+  };
+};
+
+const favouriteWeaponDuplicated = (elementToPopulate, howManyTimes) => {
+  let result = [];
+  for (let i = 0; i < howManyTimes; i++) {
+    result.push(elementToPopulate);
+  };
+  return result;
+};
```
Comment (HIPPIEKICK) [inline code/clue.js:367]: 🤩

File: code/clue.js [hunk 13]
```diff
@@ -1,59 +1,460 @@
...
+      return item;
+    };
+  });
+  mystery.weapon = mysteryWeapon;
+  mystery.favouriteRoom = mysteryWeapon.favouriteRoom;
+  weaponLoader.style.display = "none";
+  weaponCard.style.background = "#808080";
+  weaponName.innerHTML = mystery.weapon.name;
+  weaponWeight.innerHTML = `Weight: ${mystery.weapon.weight} oz`;
+  weaponImage.style.display = "inline-block";
```
Comment (HIPPIEKICK) [inline code/clue.js]: Consider adding a class instead of styling in the JS-file

File: code/clue.js [hunk 14]
```diff
@@ -1,59 +1,460 @@
...
+// Mystery storage
+
+const mystery = {
+  killer: undefined,
+  favouriteWeapon: undefined,
+  pickedWeapon: undefined,
+  weapon: undefined,
+  favouriteRoom: undefined,
+  pickedRoom: undefined,
+  room: undefined,
```
Comment (HIPPIEKICK) [inline code/clue.js]: It's even more semantic approach to initialise mystery object with null values like this :

```
const mystery = {
  killer: null,
  favouriteWeapon: null,
  pickedWeapon: null,
  weapon: null,
  favouriteRoom: null,
  pickedRoom: null,
  room: null,
};
```

**Review by HIPPIEKICK — CHANGES_REQUESTED** [review]
@MindstormingAB - Be consistent with the semicolons, either you have them everywhere or you don't have them at all.

@sofiavazs Keep it up with the constructive comments 👍

File: code/clue.js (hunk 13 above)
Comment (MindstormingAB) [reply code/clue.js]: I created a new class `.weapon-deck-clicked` and replaced `weaponCard.style.background = "#808080";` by `weaponCard.classList.add('weapon-deck-clicked');`
Was it what you meant?

**Comment by MindstormingAB** [comment]
@HIPPIEKICK  Thank you for your feedback.
I made the following changes:
- I added semicolons everywhere 😁
- I improved the if statements
- I added classes instead of styling in the JS-file
- I replaced `undefined` with `null`

**Review by HIPPIEKICK — APPROVED** [review]

**Comment by HIPPIEKICK** [comment]
> @HIPPIEKICK Thank you for your feedback.
> I made the following changes:
> 
> * I added semicolons everywhere 😁
> * I improved the if statements
> * I added classes instead of styling in the JS-file
> * I replaced `undefined` with `null`

Some semicolons missing after the arrays, no? ;)

File: code/clue.js (hunk 13 above)
Comment (HIPPIEKICK) [reply code/clue.js]: Yeah, so you can have all the styling in the CSS and then just adding classes in JS 👍

**Pull request closed without merging**